executor/index_lookup_merge_join.go

```diff
@@ -0,0 +1,760 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser/terror"
+	"github.com/pingcap/tidb/expression"
+	plannercore "github.com/pingcap/tidb/planner/core"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/memory"
+	"github.com/pingcap/tidb/util/ranger"
+	"github.com/pingcap/tidb/util/stringutil"
+	"go.uber.org/zap"
+)
+
+// IndexLookUpMergeJoin realizes IndexLookUpJoin by merge join
+type IndexLookUpMergeJoin struct {
+	baseExecutor
+
+	resultCh   <-chan *lookUpMergeJoinTask
+	cancelFunc context.CancelFunc
+	workerWg   *sync.WaitGroup
+
+	outerMergeCtx outerMergeCtx
+	innerMergeCtx innerMergeCtx
+
+	joiner      joiner
+	isOuterJoin bool
+
+	requiredRows int64
+
+	task *lookUpMergeJoinTask
+
+	indexRanges   []*ranger.Range
+	keyOff2IdxOff []int
+	innerPtrBytes [][]byte
+
+	// lastColHelper store the information for last col if there's complicated filter like col > x_col and col < x_col + 100.
+	lastColHelper *plannercore.ColWithCmpFuncManager
+
+	memTracker *memory.Tracker // track memory usage
+}
+
+type outerMergeCtx struct {
+	rowTypes      []*types.FieldType
+	joinKeys      []*expression.Column
+	keyCols       []int
+	filter        expression.CNFExprs
+	needOuterSort bool
+	compareFuncs  []expression.CompareFunc
+}
+
+type innerMergeCtx struct {
+	readerBuilder *dataReaderBuilder
+	rowTypes      []*types.FieldType
+	joinKeys      []*expression.Column
+	keyCols       []int
+	compareFuncs  []expression.CompareFunc
+	colLens       []int
+	hasPrefixCol  bool
+	desc          bool
+}
+
+type lookUpMergeJoinTask struct {
+	outerResult *chunk.Chunk
+	outerMatch  []bool
+
+	innerResult *chunk.List
+	innerIter   chunk.Iterator
+
+	sameKeyRows []chunk.Row
+	sameKeyIter chunk.Iterator
+
+	doneErr     chan error
+	done        bool
+	cursor      int
+	innerCursor int
```

Comment:
We do not need the `innerCursor`? Since we already have `innerIter`.

Reply:
Because the `innerResults` is `chunk.List`. We need a cursor to mark which chunk is under handling and an iterator to mark which row is under handling.

Reply:
check NewIterator4List